spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/DirectMessageListenerContainer.java

```diff
@@ -449,30 +449,75 @@ public void handleDelivery(String consumerTag, Envelope envelope,
 			messageProperties.setConsumerTag(consumerTag);
 			messageProperties.setConsumerQueue(this.queue);
 			Message message = new Message(body, messageProperties);
+			long deliveryTag = envelope.getDeliveryTag();
 			if (this.logger.isDebugEnabled()) {
 				this.logger.debug(this + " received " + message);
 			}
+			if (this.transactionManager != null) {
+				try {
+					if (this.isRabbitTxManager) {
+						ConsumerChannelRegistry.registerConsumerChannel(getChannel(), this.connectionFactory);
+					}
+					new TransactionTemplate(this.transactionManager, this.transactionAttribute).execute(s -> {
+						ConnectionFactoryUtils.bindResourceToTransaction(new RabbitResourceHolder(getChannel(), false),
+								this.connectionFactory, true);
+						callExecuteListener(message, deliveryTag);
+						return null;
+					});
+				}
+				finally {
+					if (this.isRabbitTxManager) {
+						ConsumerChannelRegistry.unRegisterConsumerChannel();
+					}
+				}
+			}
+			else {
+				callExecuteListener(message, deliveryTag);
+			}
+		}
+
+		private void callExecuteListener(Message message, long deliveryTag) {
 			try {
 				executeListener(getChannel(), message);
+				boolean channelLocallyTransacted = isChannelLocallyTransacted();
 				if (this.ackRequired) {
-					getChannel().basicAck(envelope.getDeliveryTag(), false);
+					if (isChannelTransacted() && !channelLocallyTransacted) {
+
+						// Not locally transacted but it is transacted so it
+						// could be synchronized with an external transaction
+						ConnectionFactoryUtils.registerDeliveryTag(this.connectionFactory, getChannel(), deliveryTag);
+
+					}
+					else {
+						getChannel().basicAck(deliveryTag, false);
+					}
+				}
+				if (channelLocallyTransacted) {
+					RabbitUtils.commitIfNecessary(getChannel());
 				}
 			}
 			catch (Exception e) {
 				this.logger.error("Failed to invoke listener", e);
-				boolean shouldRequeue = isDefaultRequeueRejected() ||
-						e instanceof MessageRejectedWhileStoppingException;
-				Throwable t = e;
-				while (shouldRequeue && t != null) {
-					if (t instanceof AmqpRejectAndDontRequeueException) {
-						shouldRequeue = false;
-					}
-					t = t.getCause();
+				rollback(deliveryTag, e);
+			}
+		}
+
+		private void rollback(long deliveryTag, Exception e) {
+			if (isChannelTransacted()) {
+				RabbitUtils.rollbackIfNecessary(getChannel());
+			}
+			if (this.ackRequired) {
+				try {
+					getChannel().basicReject(deliveryTag,
+							RabbitUtils.shouldRequeue(isDefaultRequeueRejected(), e, this.logger));
 				}
-				if (this.ackRequired) {
-					getChannel().basicNack(envelope.getDeliveryTag(), false, shouldRequeue);
+				catch (IOException e1) {
+					this.logger.error("Failed to nack message", e1);
 				}
 			}
+			if (isChannelTransacted()) {
+				RabbitUtils.commitIfNecessary(getChannel());
```

Comment:
We've already rolled back the "real" work (any publishes that were done on the channel). We do, however, have to commit the `basicReject` so the message is requeued. There's a comment (line 641 in BQC) that we should copy here.